README.md
# Manual-Screen

Десктопное GUI-приложение для расчета ручных решеток.

## Авторские права

[Zero-Clause BSD (0BSD)](https://opensource.org/licenses/0BSD)


src/CHANGELOG.md
# Лог изменений

[//]: # (YYYY-MM-DD)
[//]: # (Added, Changed, Deprecated, Removed, Fixed, Security)
[//]: # (Добавлено, Изменения, Устарело, Удалено, Исправлено, Безопасность)

## [1.3.0] - 2021-09-01

### Изменения

- Переписано на Pascal.

## [1.2.0] - 2017-11-20

### Изменения

- Интерфейс переписан на PyQt.

## [1.1.0] - 2016-07-15

### Добавлено

- Number of the profiles is added in output.
- base_gui.py extract to other repo 'dry'

## [1.0.0] - 2016-07-14

### Добавлено

- The development moved in Python.
- The size limitations made optional.
- Fixed a bug in the calculation of the number of profiles.
- Added additional checks into the calculation.


src/gui.py
# -*- coding: utf-8 -*-

# Form implementation generated from reading ui file 'gui.ui'
#
# Created by: PyQt5 UI code generator 5.9.1
#
# WARNING! All changes made in this file will be lost!

from PyQt5 import QtCore, QtGui, QtWidgets

class Ui_Dialog(object):
    def setupUi(self, Dialog):
        Dialog.setObjectName("Dialog")
        Dialog.resize(536, 185)
        self.horizontalLayout_2 = QtWidgets.QHBoxLayout(Dialog)
        self.horizontalLayout_2.setObjectName("horizontalLayout_2")
        self.verticalLayout = QtWidgets.QVBoxLayout()
        self.verticalLayout.setObjectName("verticalLayout")
        self.gridLayout = QtWidgets.QGridLayout()
        self.gridLayout.setObjectName("gridLayout")
        self.label_2 = QtWidgets.QLabel(Dialog)
        self.label_2.setObjectName("label_2")
        self.gridLayout.addWidget(self.label_2, 1, 0, 1, 1)
        self.label_3 = QtWidgets.QLabel(Dialog)
        self.label_3.setObjectName("label_3")
        self.gridLayout.addWidget(self.label_3, 2, 0, 1, 1)
        self.label = QtWidgets.QLabel(Dialog)
        self.label.setObjectName("label")
        self.gridLayout.addWidget(self.label, 0, 0, 1, 1)
        self.edt_width = QtWidgets.QLineEdit(Dialog)
        self.edt_width.setObjectName("edt_width")
        self.gridLayout.addWidget(self.edt_width, 0, 1, 1, 1)
        self.edt_length = QtWidgets.QLineEdit(Dialog)
        self.edt_length.setObjectName("edt_length")
        self.gridLayout.addWidget(self.edt_length, 1, 1, 1, 1)
        self.edt_gap = QtWidgets.QLineEdit(Dialog)
        self.edt_gap.setObjectName("edt_gap")
        self.gridLayout.addWidget(self.edt_gap, 2, 1, 1, 1)
        self.verticalLayout.addLayout(self.gridLayout)
        spacerItem = QtWidgets.QSpacerItem(20, 40, QtWidgets.QSizePolicy.Minimum, QtWidgets.QSizePolicy.Expanding)
        self.verticalLayout.addItem(spacerItem)
        self.horizontalLayout = QtWidgets.QHBoxLayout()
        self.horizontalLayout.setObjectName("horizontalLayout")
        spacerItem1 = QtWidgets.QSpacerItem(40, 20, QtWidgets.QSizePolicy.Expanding, QtWidgets.QSizePolicy.Minimum)
        self.horizontalLayout.addItem(spacerItem1)
        self.btn_run = QtWidgets.QPushButton(Dialog)
        self.btn_run.setObjectName("btn_run")
        self.horizontalLayout.addWidget(self.btn_run)
        self.verticalLayout.addLayout(self.horizontalLayout)
        self.horizontalLayout_2.addLayout(self.verticalLayout)
        self.txt_result = QtWidgets.QPlainTextEdit(Dialog)
        self.txt_result.setObjectName("txt_result")
        self.horizontalLayout_2.addWidget(self.txt_result)

        self.retranslateUi(Dialog)
        QtCore.QMetaObject.connectSlotsByName(Dialog)

    def retranslateUi(self, Dialog):
        _translate = QtCore.QCoreApplication.translate
        Dialog.setWindowTitle(_translate("Dialog", "Dialog"))
        self.label_2.setText(_translate("Dialog", "Общая длина решетки, мм"))
        self.label_3.setText(_translate("Dialog", "Прозор, мм"))
        self.label.setText(_translate("Dialog", "Ширина решетки в канале, мм"))
        self.btn_run.setText(_translate("Dialog", "Расчет"))



src/main.py
import sys
from PyQt5 import QtWidgets
sys.path.append('..')
from manifest import VERSION, DESCRIPTION
from dry.qt import msgbox, BaseMainWindow, move_cursor_to_begin, \
    get_float_number
from dry.core import InputException
import gui
from screen import Screen, LogicError, calc_screen


MAX_WIDTH = 1500
MAX_LENGTH = 2650


class MainWindow(BaseMainWindow, gui.Ui_Dialog):
    def __init__(self) -> None:
        BaseMainWindow.__init__(self, DESCRIPTION, VERSION)

    def init_widgets(self) -> None:
        pass

    def connect_actions(self) -> None:
        self.btn_run.clicked.connect(self.run)

    def clear_results(self) -> None:
        self.txt_result.clear()

    def run(self) -> None:
        self.clear_results()
        try:
            width = get_float_number(self.edt_width, (0, False), None)
            length = get_float_number(self.edt_length, (0, False), None)
            gap = get_float_number(self.edt_gap, (0, False), None)
        except InputException as err:
            msgbox(str(err))
        else:
            try:
                screen = calc_screen(width, length, gap)
            except LogicError as err:
                msgbox(str(err))
            else:
                self.output_results(screen, width, length)

    def output_results(self, screen: Screen, width: float,
                       length: float) -> None:
        self.txt_result.appendPlainText('\n'.join((
            f'Масса решетки {screen.mass:.0f} кг',
            f'Количество профиля "3999" {screen.number} шт.'
        )))
        if width > MAX_WIDTH:
            self.txt_result.appendPlainText(
                f'\nМасса занижена при ширине решетки свыше {MAX_WIDTH} мм')
        if length > MAX_LENGTH:
            self.txt_result.appendPlainText(
                f'\nМасса занижена при длине решетки свыше {MAX_LENGTH} мм')
        move_cursor_to_begin(self.txt_result)


def main() -> None:
    app = QtWidgets.QApplication(sys.argv)
    form = MainWindow()
    form.show()
    sys.exit(app.exec_())


if __name__ == '__main__':
    main()


src/screen.py
from math import ceil
from typing import Union, NamedTuple
from dry.core import Error


# Ширина профиля 3999
PROFILE_WIDTH = 9.5


class LogicError(Error):
    pass


class Screen(NamedTuple):
    number: int
    mass: float


def assert_positive(value: Union[int, float], name: str) -> None:
    if value <= 0:
        raise LogicError(
            f'Ошибка!\n{name} - значение отрицательно или равно нулю')


def calc_number_profiles(width: float, gap: float) -> int:
    result = ceil((width - gap) / (PROFILE_WIDTH + gap))
    assert_positive(result, 'Количество профилей')
    return result


def calc_mass_profile(length: float) -> float:
    result = 0.001 * length - 0.001
    assert_positive(result, 'Масса профиля')
    return result


def calc_mass_grid_canvas(length: float, number_profiles: int) -> float:
    mass_profile = calc_mass_profile(length)
    return mass_profile * number_profiles


def calc_mass_bottom_balk(length: float) -> float:
    result = 0.003 * length - 0.006
    assert_positive(result, 'Масса нижней балки')
    return result


def calc_mass_intermediate_raker(length: float) -> float:
    result = 0.001 * length - 0.022
    assert_positive(result, 'Масса перемычки')
    return result


def calc_mass_top_slide(width: float, length: float=340) -> float:
    result = 3.258e-5 * length * width + 0.00179 * length + 0.00411 * width
    assert_positive(result, 'Масса склиза')
    return result


def calc_mass_top_prop(length: float) -> float:
    result = 0.004 * length - 0.01
    assert_positive(result, 'Масса опоры')
    return result


def calc_mass_other() -> float:
    return 1.7


def calc_mass_screen(external_width: float, length: float,
                     internal_width: float,
                     number_profiles: int) -> float:
    return sum((calc_mass_bottom_balk(external_width),
                calc_mass_intermediate_raker(external_width),
                calc_mass_top_slide(internal_width),
                calc_mass_top_prop(external_width + 200),
                calc_mass_grid_canvas(length - 356, number_profiles),
                calc_mass_other()))


def calc_screen(external_width: float, length: float,
                gap: float) -> Screen:
    internal_width = external_width - 15
    number_profiles = calc_number_profiles(internal_width, gap)
    mass_screen = calc_mass_screen(external_width, length, internal_width,
                                   number_profiles)
    return Screen(
        number=number_profiles,
        mass=mass_screen)


Src/Controller.pas
unit Controller;

{$MODE OBJFPC}
{$LONGSTRINGS ON}
{$ASSERTIONS ON}
{$RANGECHECKS ON}
{$BOOLEVAL OFF}

interface

procedure Run();
procedure MainFormInit();

implementation

uses
  ProgramInfo, GuiMainForm, ScreenCalculation, GuiHelper, Measurements,
  Classes, SysUtils;

procedure MainFormInit();
begin
  MainForm.Caption := GetProgramTitle;
end;

function CreateInputData(out InputData: TInputData): string;
const
  SIncorrectValue = ' - неправильное значение.';
var
  Value: Double;
begin
  InputData := Default(TInputData);
  Result := '';

  if MainForm.EditWidth.GetRealMinEq(Value, ToMm(MinWidth)) then
    InputData.Width := Mm(Value)
  else
    Exit('Ширина решетки' + SIncorrectValue);

  if MainForm.EditLength.GetRealMinEq(Value, ToMm(MinLength)) then
    InputData.Length := Mm(Value)
  else
    Exit('Длина решетки' + SIncorrectValue);

  if MainForm.EditGap.GetRealMin(Value, 0) then
    InputData.Gap := Mm(Value)
  else
    Exit('Прозор' + SIncorrectValue);
end;

function CreateOutput(const Scr: TManualScreen): string;
var
  Lines: TStringList;
begin
  Lines := TStringList.Create;
  Lines.Add(Format('Масса решетки %.0f кг', [Scr.Weight]));
  Lines.Add(Format('Количество профиля "3999" %d шт.',
    [Scr.ProfileNumber]));
  Lines.Add(Format('Длина профиля %.0f мм', [ToMm(Scr.ProfileLength)]));
  Lines.Add('');
  if Scr.ExternalWidth > MaxWidth then
    Lines.Add(Format(
      'Масса занижена при ширине решетки свыше %.0f мм.',
      [
      ToMm(MaxWidth)]));
  if Scr.Length > MaxLength then
    Lines.Add(Format(
      'Масса занижена при длине решетки свыше %.0f мм.',
      [
      ToMm(MaxLength)]));
  Result := Lines.Text;
  Lines.Free;
end;

procedure PrintOutput(const Text: string);
begin
  MainForm.MemoOutput.Clear;
  MainForm.MemoOutput.Text := Text;
  MainForm.MemoOutput.SelStart := 0;
end;

procedure Run();
var
  InputData: TInputData;
  InputDataError: string;
  Scr: TManualScreen;
begin
  InputDataError := CreateInputData(InputData);
  if InputDataError = '' then
  begin
    CalcManualScreen(Scr, InputData);
    PrintOutput(CreateOutput(Scr));
  end
  else
    PrintOutput(InputDataError);
end;

end.


Src/GuiMainForm.pas
unit GuiMainForm;

{$MODE OBJFPC}
{$LONGSTRINGS ON}
{$ASSERTIONS ON}
{$RANGECHECKS ON}
{$BOOLEVAL OFF}

interface

uses
  Forms, StdCtrls;

type

  { TMainForm }

  TMainForm = class(TForm)
    ButtonRun: TButton;
    EditWidth: TEdit;
    EditLength: TEdit;
    EditGap: TEdit;
    Label1: TLabel;
    Label2: TLabel;
    Label3: TLabel;
    MemoOutput: TMemo;
    procedure ButtonRunClick(Sender: TObject);
    procedure FormCreate(Sender: TObject);
    procedure FormKeyPress(Sender: TObject; var Key: Char);
  private

  public

  end;

var
  MainForm: TMainForm;

implementation

uses
  Controller, LCLType;

{$R *.lfm}

{ TMainForm }

procedure TMainForm.FormKeyPress(Sender: TObject; var Key: Char);
begin
  if Key = Chr(VK_RETURN) then
    Run();
end;

procedure TMainForm.ButtonRunClick(Sender: TObject);
begin
  Run();
end;

procedure TMainForm.FormCreate(Sender: TObject);
begin
  MainFormInit();
end;

end.



Src/ScreenCalculation.pas
unit ScreenCalculation;

{$MODE OBJFPC}
{$LONGSTRINGS ON}
{$ASSERTIONS ON}
{$RANGECHECKS ON}
{$BOOLEVAL OFF}

interface

type
  TInputData = record
    Width: Double;
    Length: Double;
    Gap: Double;
  end;

  TManualScreen = record
    ExternalWidth: Double;
    Length: Double;
    Weight: Double;
    ProfileNumber: Integer;
    ProfileLength: Double;
  end;

const
  MinWidth = 0.3;
  MaxWidth = 1.5;
  MinLength = 0.5;
  MaxLength = 2.65;

procedure CalcManualScreen(out Scr: TManualScreen; const InputData: TInputData);

implementation

uses
  Math, Measurements;

const
  ProfileWidth = 0.0095;  // Ширина профиля 3999

function CalcProfileNumber(const Width, Gap: Double): Integer;
begin
  Result := Ceil((Width - Gap) / (ProfileWidth + Gap));
  Assert(Result > 0);
end;

function CalcProfileWeight(const Length: Double): Double;
begin
  Result := 1 * Length - 0.001;
  Assert(Result > 0);
end;

function CalcGridWeight(const Length: Double; const ProfileNumber: Integer): Double;
begin
  Result := CalcProfileWeight(Length) * ProfileNumber;
  Assert(Result > 0);
end;

function CalcBottomBalkWeight(const Length: Double): Double;
begin
  Result := 3 * Length - 0.006;
  Assert(Result > 0);
end;

function CalcIntermediateRakerWeight(const Length: Double): Double;
begin
  Result := 1 * Length - 0.022;
  Assert(Result > 0);
end;

function CalcTopSlideWeight(const Width, Length: Double): Double;
begin
  Result := 32.58 * Length * Width + 1.79 * Length + 4.11 * Width;
  Assert(Result > 0);
end;

function CalcTopPropWeight(const Length: Double): Double;
begin
  Result := 4 * Length - 0.01;
  Assert(Result > 0);
end;

function CalcProfileLength(const ScreenLength: Double): Double;
begin
  Result := ScreenLength - Mm(356);
  Assert(Result > 0);
end;

function CalcScreenWeight(const ExternalWidth, ProfileLength, InternalWidth: Double;
  const ProfileNumber: Integer): Double;
const
  OtherWeight = 1.7;
begin
  Result := CalcBottomBalkWeight(ExternalWidth) +
    CalcIntermediateRakerWeight(ExternalWidth) +
    CalcTopSlideWeight(InternalWidth, Mm(340)) +
    CalcTopPropWeight(ExternalWidth + Mm(200)) +
    CalcGridWeight(ProfileLength, ProfileNumber) + OtherWeight;
end;

function CalcInternalWidth(const ExternalWidth: Double): Double;
begin
  Result := ExternalWidth - Mm(15);
  Assert(Result > 0);
end;

procedure CalcManualScreen(out Scr: TManualScreen; const InputData: TInputData);
var
  InternalWidth: Double;
begin
  Scr.ExternalWidth := InputData.Width;
  Scr.Length := InputData.Length;

  InternalWidth := CalcInternalWidth(Scr.ExternalWidth);
  Scr.ProfileNumber := CalcProfileNumber(InternalWidth, InputData.Gap);
  Scr.ProfileLength := CalcProfileLength(Scr.Length);
  Scr.Weight := CalcScreenWeight(Scr.ExternalWidth, Scr.ProfileLength,
    InternalWidth, Scr.ProfileNumber);
end;

end.
